### ISSUE TITLE
(afputils) Recompiling afp/bsi/geo files for pop'n music

### ISSUE THREAD

**Issue by leeks39** [DragonMinded/bemaniutils#85]
I have been tinkering with character animations in pop'n music by using afputils and noticed there is no current way to recompile any of the afp/bsi/geo files. I did read that the format is quite complicated, but I wanted to see if it would ever be possible to have a way to modify the bytecode and convert it back to its original state for use in-game.

I apologize if I'm missing information for this request or am putting it in the wrong spot.

**Comment by DragonMinded** [comment]
The format is insanely complicated. It's basically a fork of flash, with a bunch of Konami-specific additions and a completely new container format, as well as a second layer container (either the TXP2 or IFS file that holds the AFP/BSI/GEO and textures) on top of AFP itself. Unpacking/repacking the second layer is a solved problem. Mon's stuff can repack IFS, and my own code can repack TXP2. The AFP side of things is another thing altogether. I think at some point it would be possible/desirable to repack AFP files (BSI are just byteswap scrambling files, so as long as the instructions encode for the same length as the AFP file they go with, it's fully reversible). Both AFP and GEO files have bits and chunks that aren't quite understood, but for Pop'n it is basically 100% understood what every bit does. You wouldn't really need to mess with the bytecode though, for Pop'n that's only used for notifying the game of specific frames and for layer masks.

Authoring tools for new animations would be difficult, as it would basically mean re-creating the flash toolkit. I don't know how Konami makes the files internally. Script-based might work, but again, lots of work. Honestly it's not really something I've had the need for, but I know one of the Pop'n community's holy grails has been the ability to add characters and animations. I'm probably not the one to do it, however, as I've lost interest in the format.

**Comment by jorpz** [comment]
Yes, adding CS characters (and custom **I GUESS**) has been a longtime dream project for some people, and at one point someone (maybe even leeks39 here is that someone!) was working on a tool that could port CS characters over. Not sure what came of that, but assuming that goes somewhere it would be interesting to see how broad they're going. If it's JUST to convert CS popn characters to the new format that's great, but that's like if we only had a tool that did something like "oh my tool can read DDR SSQ files but only the lighting data!" That's a good start, but only a start. I'm not sure what the end goal of something like that would be, I mean the pop'n music characters obviously would be the end goal for that project, but what else would be needed? AFAIK custom jubeat markers are already being made, so what other stuff? Other than just knowledge?

tldr someone somewhere was working on a tool to port the old popn characters and if they do that code might provide some more insight into the afp format but I don't know what help that is. Thanks for coming to my tedtalk.

**Comment by ameeho67k** [comment]
Given the complexity of the TXP2/IFS container layers mentioned, has anyone mapped out the specific header offsets or checksum calculations required to maintain file integrity during a repack? Focusing on a partial re-serialization of the bytecode before addressing the full container structure might be a more viable way to test if the game engine accepts modified AFP blobs.
